daemon/compartment.c

```diff
@@ -600,11 +628,64 @@ compartment_cleanup(compartment_t *compartment, bool is_rebooting)
 		compartment->start_timer = NULL;
 	}
 
+	for (list_t *l = compartment->module_instance_list; l; l = l->next) {
+		compartment_module_instance_t *c_mod = l->data;
+		compartment_module_t *module = c_mod->module;
+		if (NULL == module->cleanup)
+			continue;
+
+		module->cleanup(c_mod->instance, is_rebooting);
+	}
+
+	compartment->pid = -1;
+	compartment->pid_early = -1;
+
+	// check if some cleanup routines registered any helper childs
+	// and wait for them in sig child handler
+	if (compartment->helper_child_list) {
+		compartment->is_doing_cleanup = true;
+		return;
+	}
+
 	compartment_state_t state =
 		is_rebooting ? COMPARTMENT_STATE_REBOOTING : COMPARTMENT_STATE_STOPPED;
 	compartment_set_state(compartment, state);
 }
 
+void
+compartment_sigchld_handle_helpers(compartment_t *compartment, event_signal_t *sig)
+{
+	int status = 0;
+
+	list_t *still_active = NULL;
+	for (list_t *l = compartment->helper_child_list; l; l = l->next) {
+		compartment_helper_child_t *child = l->data;
+		pid_t pid = waitpid(child->pid, &status, WNOHANG);
+		if (pid == child->pid) {
+			DEBUG("Reaped helper child %s (pid=%d) for compartment %s", child->name,
+			      pid, compartment_get_description(compartment));
+			compartment_helper_child_free(child);
+		} else {
+			still_active = list_append(still_active, child);
+		}
+	}
+	list_delete(compartment->helper_child_list);
+	compartment->helper_child_list = still_active;
+
+	if (compartment_is_cleanup_done(compartment)) {
+		DEBUG("CLEANUP DONE, all pending helpers reaped!");
+		/* remove the sigchld callback for this compartment from the event loop */
+		event_remove_signal(sig);
+		event_signal_free(sig);
+		compartment->is_doing_cleanup = false;
+		compartment_state_t state = compartment->is_rebooting ?
+						    COMPARTMENT_STATE_REBOOTING :
+						    COMPARTMENT_STATE_STOPPED;
+		compartment_set_state(compartment, state);
+		compartment->is_rebooting = false;
```

Comment:
would is_rebooting be mergable with  COMPARTMENT_STATE_REBOOTING?

Reply:
nop since we have to wait for child before setting the state